src/System.Net.Security/src/System/Net/Security/SniHelper.cs

```diff
@@ -0,0 +1,391 @@
+            //   insensitively.
+
+            string idnEncodedString;
+            try
+            {
+                idnEncodedString = s_encoding.GetString(bytes);
+            }
+            catch (DecoderFallbackException)
+            {
+                return null;
+            }
+
+            try
+            {
+                return s_idnMapping.GetUnicode(idnEncodedString);
+            }
+            catch (ArgumentException)
+            {
+                // client has not done IDN mapping
+                return idnEncodedString;
+            }
+        }
+
+        private static int ReadUInt24BigEndian(ReadOnlySpan<byte> bytes)
+        {
+            return (bytes[0] << 16) | (bytes[1] << 8) | bytes[2];
+        }
+
+        private static ReadOnlySpan<byte> SkipBytes(ReadOnlySpan<byte> bytes, int numberOfBytesToSkip)
+        {
+            return (numberOfBytesToSkip < bytes.Length) ? bytes.Slice(numberOfBytesToSkip) : ReadOnlySpan<byte>.Empty;
+        }
+
+        // Opaque type is of structure:
+        //   - length (minimum number of bytes to hold the max value)
+        //   - data (length bytes)
+        // We will only use opaque types which are of max size: 255 (length = 1) or 2^16-1 (length = 2).
+        // We will call them SkipOpaqueType`length`
+        private static ReadOnlySpan<byte> SkipOpaqueType1(ReadOnlySpan<byte> bytes)
+        {
+            const int OpaqueTypeLengthSize = sizeof(byte);
+            if (bytes.Length < OpaqueTypeLengthSize)
+            {
+                return ReadOnlySpan<byte>.Empty;
+            }
+
+            byte length = bytes[0];
+            int totalBytes = OpaqueTypeLengthSize + length;
+
+            return SkipBytes(bytes, totalBytes);
+        }
+
+        private static ReadOnlySpan<byte> SkipOpaqueType2(ReadOnlySpan<byte> bytes, out bool invalid)
+        {
+            const int OpaqueTypeLengthSize = sizeof(ushort);
+            if (bytes.Length < OpaqueTypeLengthSize)
+            {
+                invalid = true;
+                return ReadOnlySpan<byte>.Empty;
+            }
+
+            ushort length = BinaryPrimitives.ReadUInt16BigEndian(bytes);
+            int totalBytes = OpaqueTypeLengthSize + length;
+
+            invalid = bytes.Length < totalBytes;
+            if (invalid)
+            {
+                return ReadOnlySpan<byte>.Empty;
+            }
+            else
+            {
+                return bytes.Slice(totalBytes);
+            }
+        }
+
+        private static IdnMapping CreateIdnMapping()
+        {
+            return new IdnMapping()
+            {
+                // Per spec "AllowUnassigned flag MUST be set". See comment above GetSniFromServerNameList for more details.
+                AllowUnassigned = true
+            };
+        }
+
+        private static Encoding CreateEncoding()
+        {
+            return Encoding.GetEncoding("utf-8", new EncoderExceptionFallback(), new DecoderExceptionFallback());
```

Comment:
Why does this need to use GetEncoding("utf-8", ...) rather than new UTF8Encoding() or Encoding.UTF8?

Reply:
@stephentoub Utf8Encoding is in different assembly and it would introduce new dependency (System.Text.Encoding.Extensions) where we can create this exact type without introducing it by doing above

Reply:
Is that a dependency we care about?  The implementation of UTF8Encoding is in corelib.  And using Encoding.GetEncoding(string) here would seem to a) be more expensive, and b) not necessarily produce a UTF8Encoding as (if I'm understanding the code correctly) it'll go through any registered providers before falling back to looking up what encoder should be used by internal tables.

(And I assume Encoding.UTF8 can't be used because the fallbacks it needs are different?)

cc: @tarekgh, @weshaggard 

Anyway, don't block on this one.